src/merkle/store/tests.rs

```diff
@@ -647,6 +647,70 @@ fn test_constructors() -> Result<(), MerkleError> {
     Ok(())
 }
 
+#[test]
+fn node_path_should_be_truncated_by_midtier_insert() {
+    let key = 0b11010010_11001100_11001100_11001100_11001100_11001100_11001100_11001100_u64;
+
+    let mut store = MerkleStore::new();
+    let root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+
+    // insert first node - works as expected
+    let depth = 64;
+    let node = [Felt::new(key); WORD_SIZE];
+    let index = NodeIndex::new(depth, key).unwrap();
+    let root = store.set_node(root, index, node).unwrap().root;
+    let result = store.get_node(root, index).unwrap();
+    let path = store.get_path(root, index).unwrap().path;
+    assert_eq!(node, result);
+    assert_eq!(path.depth(), depth);
+    assert!(path.verify(index.value(), result, &root));
+
+    // flip the first bit of the key and insert the second node on a different depth
+    let key = key ^ (1 << 63);
+    let key = key >> 8;
+    let depth = 56;
+    let node = [Felt::new(key); WORD_SIZE];
+    let index = NodeIndex::new(depth, key).unwrap();
+    let root = store.set_node(root, index, node).unwrap().root;
+    let result = store.get_node(root, index).unwrap();
+    let path = store.get_path(root, index).unwrap().path;
+    assert_eq!(node, result);
+    assert_eq!(path.depth(), depth);
+    assert!(path.verify(index.value(), result, &root));
+
+    // attempt to fetch a path of the second node to depth 64
+    // should fail because the previously inserted node will remove its sub-tree from the set
+    let key = key << 8;
+    let index = NodeIndex::new(64, key).unwrap();
+    assert!(store.get_node(root, index).is_err());
+}
+
+#[test]
+fn get_leaf_depth_works() {
+    let key = u64::MAX;
+
+    let mut store = MerkleStore::new();
+    let mut root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+
+    // use the same root to test different paths
+    for depth in (1..=64).rev() {
+        let mask = 64 - depth as u64;
+        let key = (key << mask) >> mask;
+        let node = [Felt::new(key); WORD_SIZE];
+        let index = NodeIndex::new(depth, key >> mask).unwrap();
```

Comment:
question: is it just me that doesn't like this API? This seems quite error prone to me.

Reply:
btw, isn't this zero most of the time? here is how I'm reading this:

- we have a new binding to `key` on line `698` that remove the `64 - depth` highbits from the key
- then here we remove the `64 - depth` lowbits of the key

the two operations above together, for as long as `depth<32`, will drop all bits and we end up with zero, no? For `depth>32` we are left with the middle bits, which is always `1` because the original `key` is `u64::MAX`, so it seems that for depths from 0 to 31 position 0 is always used, the from position 32 onwards the position `2**n-1` are used, meaning the node goes to a different leaf, was that the idea? maybe we can make the intent of the test a bit easier to follow?

Reply:
> question: is it just me that doesn't like this API? This seems quite error prone to me.

I don't particularly like it either - but it is also not the "intended" use of the API - we are using it like this for testing. The intended use is that the provided depth/value pair are internally consistent - but how the user derives them is up to the user.

We could add another constructor which would do this computation internally. Maybe something like `new_truncated()` or something like that - but I'm not sure that'll be much more clear (i.e., the user would basically provide a depth/value pair which is not internally consistent and would ask the constructor to fix it.

Reply:
> The intended use is that the provided depth/value pair are internally consistent - but how the user derives them is up to the user.

By internally consistent I assume you mean the depth can change, but the walk on the structure has to be the same. So if we start with `[left, left, left, right]`, and change the depth to 2, we must end with `[left, left]` and not `[left, right]`. `NodeIndex` has that when using `move_up`, but the walk needs it for `move_down`, if we change that the user doesn't have to do any bit manipulation. I think we don't use `move_up` consistency for anything.

Reply:
I don't mind adding `move_down()` method here. What I was referring to is not adding extra logic to the constructor.

Reply:
I was trying to say that we don't have to add an extra constructor, or require the user to do bit shifting if we rename `move_up` to `move_down` and change the code to match accordingly.